File: CommunitySurvey/Services/AuthService.swift
import Foundation

protocol AuthServiceProtocol: Sendable {
    func register(
        fullName: String,
        fathersName: String,
        gender: String,
        mobile: String,
        aadhaar: String,
        address: String,
        role: UserRole,
        organizationId: String?,
        organizationName: String?,
        organizationType: String?,
        organizationCode: String?,
        state: String,
        district: String,
        pincode: String,
        education: String,
        occupation: String,
        socialCategory: String,
        city: String
    ) async throws -> AuthSession
    func login(mobile: String) async throws -> User
    func getProfile() async throws -> User
    func updateProfile(fullName: String?) async throws -> User
    func logout() async
    func isAuthenticated() -> Bool
    func joinOrganization(code: String) async throws -> User

    func requestOTP(mobileNumber: String, countryCode: String) async throws -> OTPResponse
    func verifyOTP(transactionID: String, otp: String, mobileNumber: String, countryCode: String) async throws -> AuthSession
    func refresh(session: AuthSession) async throws -> AuthSession
}

@MainActor
struct AuthService: AuthServiceProtocol {
    private let apiClient: APIClientProtocol
    private let keychainService: KeychainService

    init(apiClient: APIClientProtocol, keychainService: KeychainService) {
        self.apiClient = apiClient
        self.keychainService = keychainService
    }

    func register(
        fullName: String,
        fathersName: String,
        gender: String,
        mobile: String,
        aadhaar: String,
        address: String,
        role: UserRole,
        organizationId: String?,
        organizationName: String?,
        organizationType: String?,
        organizationCode: String?,
        state: String,
        district: String,
        pincode: String,
        education: String,
        occupation: String,
        socialCategory: String,
        city: String
    ) async throws -> AuthSession {
        let response: RegisterUserResponse = try await apiClient.request(
            path: "/auth/register",
            method: .post,
            body: RegisterUserRequest(
                fullName: fullName,
                fathersName: fathersName,
                gender: gender,
                mobile: mobile,
                aadhaar: aadhaar,
                address: address,
                role: role,
                organizationId: organizationId,
                organizationName: organizationName,
                organizationType: organizationType,
                organizationCode: organizationCode,
                state: state,
                district: district,
                pincode: pincode,
                education: education,
                occupation: occupation,
                socialCategory: socialCategory,
                city: city
            ),
            requiresAuthentication: false,
            responseType: RegisterUserResponse.self
        )
        try keychainService.save(token: response.token)
        return makeSession(token: response.token, user: response.user, mobileNumber: mobile, countryCode: "+91")
    }

    func login(mobile: String) async throws -> User {
        let response: LoginUserResponse = try await apiClient.request(
            path: "/auth/login",
            method: .post,
            body: LoginUserRequest(mobile: mobile),
            requiresAuthentication: false,
            responseType: LoginUserResponse.self
        )
        try keychainService.save(token: response.token)
        return response.user
    }

    func getProfile() async throws -> User {
        let response: ProfileResponse = try await apiClient.request(
            path: "/auth/profile",
            method: .get,
            body: Optional<EmptyRequest>.none,
            requiresAuthentication: true,
            responseType: ProfileResponse.self
        )
        return response.user
    }

    func updateProfile(fullName: String?) async throws -> User {
        let response: ProfileResponse = try await apiClient.request(
            path: "/auth/profile",
            method: .put,
            body: UpdateProfileRequest(fullName: fullName),
            requiresAuthentication: true,
            responseType: ProfileResponse.self
        )
        return response.user
    }

    func logout() async {
        try? keychainService.deleteToken()
    }

    func isAuthenticated() -> Bool {
        keychainService.getToken() != nil
    }

    func joinOrganization(code: String) async throws -> User {
        struct JoinOrgRequest: Encodable {
            let organizationCode: String
        }
        struct JoinOrgResponse: Decodable {
            let success: Bool
            let user: User
        }
        let response = try await apiClient.request(
            path: "/auth/join-org",
            method: .post,
            body: JoinOrgRequest(organizationCode: code),
            requiresAuthentication: true,
            responseType: JoinOrgResponse.self
        )
        return response.user
    }

    func requestOTP(mobileNumber: String, countryCode: String) async throws -> OTPResponse {
        let response: OTPLoginResponse = try await apiClient.request(
            path: "/auth/send-otp",
            method: .post,
            body: LoginUserRequest(mobile: mobileNumber),
            requiresAuthentication: false,
            responseType: OTPLoginResponse.self
        )
        return OTPResponse(transactionID: UUID().uuidString, expiresIn: 60, otp: response.otp)
    }

    func verifyOTP(transactionID: String, otp: String, mobileNumber: String, countryCode: String) async throws -> AuthSession {
        let response: VerifyOTPResponse = try await apiClient.request(
            path: "/auth/verify-otp",
            method: .post,
            body: VerifyOTPRequest(mobile: mobileNumber, otp: otp),
            requiresAuthentication: false,
            responseType: VerifyOTPResponse.self
        )
        try keychainService.save(token: response.token)
        return makeSession(token: response.token, user: response.user, mobileNumber: mobileNumber, countryCode: countryCode)
    }

    func refresh(session: AuthSession) async throws -> AuthSession { session }

    private func makeSession(token: String, user: User, mobileNumber: String, countryCode: String) -> AuthSession {
        AuthSession(
            accessToken: token,
            refreshToken: "",
            expiresAt: Date().addingTimeInterval(3600),
            user: AuthenticatedUser(
                id: user.id,
                mobileNumber: user.mobile ?? mobileNumber,
                countryCode: countryCode,
                role: user.role,
                organization: user.organizationId
            )
        )
    }
}

struct EmptyRequest: Encodable, Sendable { }

File: flutter_app/lib/services/auth_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:community_survey/core/network/api_client.dart';
import 'package:community_survey/core/network/api_endpoints.dart';
import 'package:community_survey/models/auth_session.dart';
import 'package:community_survey/models/user.dart';

class AuthService {
  final Dio _dio;

  AuthService(this._dio);

  Future<AuthSession> login({required String mobile, required String password}) async {
    final response = await _dio.post(
      ApiEndpoints.login,
      data: {
        'mobile': mobile,
        'password': password,
      },
    );
    final data = response.data;
    if (data['success'] == true) {
      return AuthSession.fromJson(data as Map<String, dynamic>);
    } else {
      throw Exception(data['message'] ?? 'Login failed.');
    }
  }

  Future<AuthSession> register({
    required String fullName,
    String fathersName = '',
    required String gender,
    required String mobile,
    required String password,
    String aadhaar = '',
    required String address,
    String? organizationId,
    String? organizationName,
    String? organizationType,
    String? organizationCode,
    required String state,
    required String district,
    required String pincode,
    required String education,
    required String occupation,
    required String socialCategory,
    required String city,
  }) async {
    final response = await _dio.post(
      ApiEndpoints.register,
      data: {
        'fullName': fullName,
        'fathersName': fathersName,
        'gender': gender,
        'mobile': mobile,
        'password': password,
        if (aadhaar.isNotEmpty) 'aadhaar': aadhaar,
        'address': address,
        if (organizationId != null) 'organizationId': organizationId,
        if (organizationName != null) 'organizationName': organizationName,
        if (organizationType != null) 'organizationType': organizationType,
        if (organizationCode != null) 'organizationCode': organizationCode,
        'state': state,
        'district': district,
        'pincode': pincode,
        'education': education,
        'occupation': occupation,
        'socialCategory': socialCategory,
        'city': city,
      },
    );
    final data = response.data;
    if (data['success'] == true) {
      return AuthSession.fromJson(data as Map<String, dynamic>);
    } else {
      throw Exception(data['message'] ?? 'Registration failed.');
    }
  }

  Future<User> updateProfile({
    String? fullName,
    String? gender,
    String? aadhaar,
    String? address,
    String? state,
    String? district,
    String? city,
    String? pincode,
    String? education,
    String? occupation,
    String? socialCategory,
  }) async {
    final response = await _dio.put(
      ApiEndpoints.profile,
      data: {
        if (fullName != null) 'fullName': fullName,
        if (gender != null) 'gender': gender,
        if (aadhaar != null) 'aadhaar': aadhaar,
        if (address != null) 'address': address,
        if (state != null) 'state': state,
        if (district != null) 'district': district,
        if (city != null) 'city': city,
        if (pincode != null) 'pincode': pincode,
        if (education != null) 'education': education,
        if (occupation != null) 'occupation': occupation,
        if (socialCategory != null) 'socialCategory': socialCategory,
      },
    );
    final data = response.data;
    if (data['success'] == true) {
      return User.fromJson(data['user'] as Map<String, dynamic>);
    } else {
      throw Exception(data['message'] ?? 'Failed to update profile.');
    }
  }

  Future<User> getProfile() async {
    final response = await _dio.get(ApiEndpoints.profile);
    final data = response.data;
    if (data['success'] == true) {
      return User.fromJson(data['user'] as Map<String, dynamic>);
    } else {
      throw Exception(data['message'] ?? 'Failed to load profile.');
    }
  }

  Future<User> joinOrganization(String code) async {
    final response = await _dio.post(
      ApiEndpoints.joinOrg,
      data: {'organizationCode': code},
    );
    final data = response.data;
    if (data['success'] == true) {
      return User.fromJson(data['user'] as Map<String, dynamic>);
    } else {
      throw Exception(data['message'] ?? 'Failed to join group.');
    }
  }
}

final authServiceProvider = Provider<AuthService>((ref) {
  return AuthService(ref.watch(dioProvider));
});
